src/main/java/io/unlaunch/store/UnlaunchHttpDataStore.java

```diff
@@ -27,81 +28,144 @@
 /**
  * This class is responsible for fetching feature flags using the HTTP API and maintain a {@link ConcurrentHashMap} of
  * feature flags that are downloaded from the server in memory.
- *
+ * <p>
  * This class is not mutable and hence it is thread-safe.
- *
+ * <p>
  * It is safe to call run this class periodically and hence can be used by {@link RefreshableDataStoreProvider}s.
- *
  */
 final class UnlaunchHttpDataStore implements UnlaunchDataStore, Runnable {
 
     private final AtomicReference<Map<String, FeatureFlag>> refFlagsMap;
     private AtomicReference<String> projectNameRef = new AtomicReference<>();
     private AtomicReference<String> environmentNameRef = new AtomicReference<>();
     private final UnlaunchRestWrapper restWrapper;
+    private final UnlaunchGenericRestWrapper s3BucketRestWrapper;
     private final CountDownLatch gate;
     private final JsonObjectConversionHelper flagService = new JsonObjectConversionHelper();
     private final JSONParser parser = new JSONParser();
     private final AtomicBoolean initialSyncSuccessful;
+    private final AtomicBoolean sync0Complete = new AtomicBoolean(false);
     private final AtomicInteger numHttpCalls = new AtomicInteger(0);
 
     private static final Logger logger = LoggerFactory.getLogger(UnlaunchHttpDataStore.class);
 
-    protected UnlaunchHttpDataStore(UnlaunchRestWrapper restWrapper, CountDownLatch gate, AtomicBoolean initialSyncSuccessful) {
+    protected UnlaunchHttpDataStore(UnlaunchRestWrapper restWrapper, UnlaunchGenericRestWrapper s3BucketRestWrapper,
+                                    CountDownLatch gate, AtomicBoolean initialSyncSuccessful) {
         this.restWrapper = restWrapper;
+        this.s3BucketRestWrapper = s3BucketRestWrapper;
         this.gate = gate;
         this.initialSyncSuccessful = initialSyncSuccessful;
         this.refFlagsMap = new AtomicReference<>(new HashMap<>());
     }
 
-    @Override
-    public void run() {
-        numHttpCalls.incrementAndGet();
-        Object obj = null;
+    private boolean sync0()  {
 
         try {
-            Response response = restWrapper.get();
-            if (response.getStatus() == 304) {
-                logger.debug("synced flags with the server. No update. In memory data store has {} flags",
-                        refFlagsMap.get().size());
-            } else if (response.getStatus() == 200) {
-                obj = parser.parse(response.readEntity(String.class));
-                JSONObject resBodyJson = (JSONObject) obj;
+            Response response = s3BucketRestWrapper.get();
 
-                JSONObject data = (JSONObject) resBodyJson.get("data");
-                projectNameRef.set((String)data.get("projectName"));
-                environmentNameRef.set((String)data.get("envName"));
-                JSONArray flags = (JSONArray) data.get("flags");
+            // If the s3 object doesn't exist, attempt a regular sync
+            if (response.getStatus() != 200) {
```

Comment:
If I'm new user, new flag is not created yet. When calling S3, it would return something not 200 right? If yes, then that's good, else line 80 if (numFlags > 1) would not work.

Reply:
Fixed the condition to be > 0